### ISSUE TITLE
add completions for Cloud Foundry CLI

### ISSUE THREAD

**Issue by dkoper** [vladimir-kotikov/clink-completions#78]
[Cloud Foundry](docs.cloudfoundry.org) is an open source, multi cloud application platform and it has a [Command Line tool](https://github.com/cloudfoundry/cli) to interact with it.

The `cf` CLI supports bash completion "natively" as offered by the [command line parser library](https://godoc.org/github.com/jessevdk/go-flags) it uses. Enabling it is as simple as installing the following bash completion file:
```
# bash completion for cf

_cf() {
    # All arguments except the first one
    args=("${COMP_WORDS[@]:1:$COMP_CWORD}")

    # Only split on newlines
    local IFS=$'\n'

    # Call completion (note that the first element of COMP_WORDS is
    # the executable itself)
    COMPREPLY=($(GO_FLAGS_COMPLETION=1 ${COMP_WORDS[0]} "${args[@]}"))
    return 0
}

complete -F _cf cf
```

In other words, running the `cf` command with environment variable `GO_FLAGS_COMPLETION=1` makes it return a list of completions (subcommands and flags).

If we could have a clink completion script for that, it may be fairly easy to support completions for this command, as well as any other Goflags library based CLI.

**Comment by chrisant996** [comment]
I would like include this, but it has a dangerous problem:

It runs the entire command line on every completion, including other commands in the command line:
`del * & cf `<kbd>Tab</kbd> will actually run both the `del *` and the `cf` command.

This can be partially mitigated by using `clink.parseline()` and using only the last `line_state`.

But that's not sufficient, either -- for example:
`>foo cf `<kbd>Tab</kbd> will run the command and redirect output to `foo`.

To be able to include this, the argmatcher would need to be change to be more discriminating about what it runs.  For example, it could use [Responding to Arguments in Argmatchers](https://chrisant996.github.io/clink/clink.html#responsive-argmatchers) and `user_data` to collect the words that are relevant to the `cf` command itself, and run only `cf` with only those arguments.

**Comment by chrisant996** [comment]
Another reason it's dangerous is because if there's a `cf.exe` installed that isn't Cloud Foundry, then the argmatcher will _run the other program with the full command line_ in an attempt to find completions for Cloud Foundry.  And that's not going to work as intended.

(This isn't just hypothetical -- on my work computers there's an internal program `cf` that isn't Cloud Foundry.)

I'll see if there's a reasonable way to confirm that the `cf` program is indeed Cloud Foundry -- hopefully Cloud Foundry includes a version resource, in which case [os.getfileversion()](https://chrisant996.github.io/clink/clink.html#os.getfileversion) could be used to verify the program.

**Comment by chrisant996** [comment]
Completions for Cloud Foundry CLI have been added in commit f334fbad6593a99a7b366b05abf6adf5edc484b5.
